pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,737 @@
+			reqLogger.Info(fmt.Sprintf("%s already done, skip", key))
+			continue
+		}
+		result, err := c[key](cl, upgradeConfig, reqLogger)
+
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("error when %s", key))
+			condition.Reason = fmt.Sprintf("%s not done", key)
+			condition.Message = err.Error()
+			conditions.SetCondition(*condition)
+			history.Conditions = conditions
+			upgradeConfig.Status.History.SetHistory(*history)
+			err = cl.Status().Update(context.TODO(), upgradeConfig)
+			if err != nil {
+				return err
+			}
+			return err
+		}
+		if result {
+			condition.CompleteTime = &metav1.Time{Time: time.Now()}
+			condition.Reason = fmt.Sprintf("%s succeed", key)
+			condition.Message = fmt.Sprintf("%s succeed", key)
+			condition.Status = corev1.ConditionTrue
+			conditions.SetCondition(*condition)
+			history.Conditions = conditions
+			upgradeConfig.Status.History.SetHistory(*history)
+			err = cl.Status().Update(context.TODO(), upgradeConfig)
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Info(fmt.Sprintf("%s not done, skip following steps", key))
+			condition.Reason = fmt.Sprintf("%s not done", key)
+			condition.Message = fmt.Sprintf("%s still in progress", key)
+			conditions.SetCondition(*condition)
+			history.Conditions = conditions
+			upgradeConfig.Status.History.SetHistory(*history)
+			err = cl.Status().Update(context.TODO(), upgradeConfig)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+	history.Phase = upgradev1alpha1.UpgradePhaseUpgraded
+	history.CompleteTime = &metav1.Time{Time: time.Now()}
+	upgradeConfig.Status.History.SetHistory(*history)
+	err = cl.Status().Update(context.TODO(), upgradeConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+// check several things about the cluster and report problems
+// * critical alerts
+// * degraded operators (if there are critical alerts only)
+func performClusterHealthCheck(c client.Client) (bool, error) {
+	sa := &corev1.ServiceAccount{}
+
+	err := c.Get(context.TODO(), types.NamespacedName{"openshift-monitoring", "prometheus-k8s"}, sa)
+	if err != nil {
+		return false, err
+	}
+
+	tokenSecret := ""
+	for _, secret := range sa.Secrets {
+		if strings.HasPrefix(secret.Name, "prometheus-k8s-token") {
+			tokenSecret = secret.Name
+		}
+	}
+	if len(tokenSecret) == 0 {
+		return false, fmt.Errorf("failed to find token secret for prommetheus-k8s SA")
+	}
+
+	log.Info(fmt.Sprintf("found out secret %s", tokenSecret))
+
+	secret := &corev1.Secret{}
+
+	err = c.Get(context.TODO(), types.NamespacedName{"openshift-monitoring", tokenSecret}, secret)
+	if err != nil {
+		return false, err
+	}
+
+	token := secret.Data[corev1.ServiceAccountTokenKey]
+
+	route := &routev1.Route{}
+	err = c.Get(context.TODO(), types.NamespacedName{Namespace: "openshift-monitoring", Name: "prometheus-k8s"}, route)
+	if err != nil {
+		return false, err
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```

Comment:
I see the docs for this say:
```
    // InsecureSkipVerify controls whether a client verifies the
    // server's certificate chain and host name.
    // If InsecureSkipVerify is true, TLS accepts any certificate
    // presented by the server and any host name in that certificate.
    // In this mode, TLS is susceptible to man-in-the-middle attacks.
    // This should be used only for testing.
    InsecureSkipVerify bool
```

Is this left over from testing?

Reply:
Cluster installed in `int` env, are using self signed ca, see comments here: https://github.com/openshift/managed-upgrade-operator/pull/9#discussion_r432184617

Reply:
https://issues.redhat.com/browse/OSD-3945 created as a followup